spock-core/src/main/java/org/spockframework/runtime/ParameterizedFeatureNode.java

```diff
@@ -33,9 +37,38 @@ public SpockExecutionContext execute(SpockExecutionContext context, DynamicTestE
     return context;
   }
 
+  @Override
+  public void after(SpockExecutionContext context) throws Exception {
+    List<Throwable> failures = context.getFailures();
+    if (!failures.isEmpty()) {
+      Throwable failure;
+      if (failures.size() == 1) {
+        failure = failures.get(0);
+      } else {
+        failure = new MultipleFailuresError(null, failures);
```

Comment:
This basically repeats the failures from the iterations in its parent?

Reply:
the idea is:
rolled & one iteration failed => error of the iteration is rethrown for the feature
rolled & multiple iterations failed => `MultipleFailuresError` with the errors of the iterations is thrown
unrolled => SpockAssertionError is thrown with the iteration errors as cause

Reply:
Like this, but ignore the names, for this run there were **no** `@PendingFeature` annotations present:
![grafik](https://user-images.githubusercontent.com/325196/75081708-717bd500-5510-11ea-8d0a-bdee3185d039.png)

Reply:
I don't really see the benefit, the result is already reported for the individual iterations. As long as there was no error with the data provider the container should be green. And IMHO it leaks stuff from the `PendingFeatureExtension` into core logic.
If we rollup, then it would exit on the first failure instead of continuing the execution collecting all the results. As you indicated you are not interested in the individual results anyway.

Reply:
> I don't really see the benefit, the result is already reported for the individual iterations.

Well, the intention was, that if some tool or report shows only the feature (even if unrolled), that you can immediately see all errors of the child iterations.

> As long as there was no error with the data provider the container should be green.

Do you think so? I don't really agree. If something is wrong with an iteration, the feature shouldn't be green imho, as the feature was not successfully tested.

> And IMHO it leaks stuff from the PendingFeatureExtension into core logic.

I don't think so. The pending feature extension reverts the outcome, that has nothing to do with showing on the feature whether there was some problem with some iteration.

And besides that, especially with pending feature, we need to fail the feature if all iterations were green. For anything else, we would need to first run all iterations and capture their result and events and then fail or forward if all were green or not (with unrolled feature that is).

> If we rollup, then it would exit on the first failure instead of continuing the execution collecting all the results. As you indicated you are not interested in the individual results anyway.

Where did I indicate that? I do not agree here actually, as doesn't the documentation.
The documentation says that all iterations are run and that all failing iterations are reported.

Quote from http://spockframework.org/spock/docs/1.3/all_in_one.html#data-tables:
> If an iteration fails, the remaining iterations will nevertheless be executed. All failures will be reported.

Actually this is currently not the case, but either it was the case previously or it was at least intended. Doing it like I did it for rolledup features adjusts the logic to the documentation and shows all failures.

Imho we should not cancel execution on first failure, but report all failures as indicated by the documentation, because I want to know all iterations that failed and not fix the failing one, rerun, fix the next failing one, rerun until all is green.

Reply:
And a bit later in the docs it also says `Note that unrolling has no effect on how the method gets executed; it is only an alternation in reporting.` :-)

Reply:
>>  As you indicated you are not interested in the individual results anyway.

> Where did I indicate that?

_You_ in this case means the user that attached `@Rollup`.

Reply:
Ah, I see, but me - if using `@Rollup` - would rather want to say "I have a data driven feature with 5000 calculated iterations, I don't want them all reported individually and increase the test count, but if some are failing, I want to know which and I want them all", which seems also being the original intent as I interpret the docs. :-)

Reply:
As of right now, we are still waiting for https://github.com/junit-team/junit5/issues/2188 to implement `@Rollup` with that feature we would also gain access to the result without having to save it in the `IterationNode`. I would wait with the integration of this pr until we get that.

Reply:
Actually I have the rollup locally already, I've set the listener via reflection and it works fine. :-)

But even with that PR, at least from what Marc suggested, you need to store the results somewhere, at least if you want to aggregate in the parent, because you can supply a custom listener and that listener has to store the results somewhere. And imho the iteration node is the proper place, don't you think?

Independent from that PR, it would be nice if we come to a conclusion here when to show what where, because I then can already adapt this here and maybe even finish 99% of the rollup. Or did that statement mean I convinced you already? :-D If so, then how do you think the PR would or should change the implementation here, where should the listener store the result instead?

Reply:
I'd create a new instance of a collecting listener, similar to what JUnit does with the `ThrowableCollector` and check the result of the listener afterwards. I don't really think that the Node is the proper place to store the result. Currently you loose information if an extension decides to just re-execute the IterationNode.

Reply:
Oh, can it?
I mean in a way that the listener is called multiple times?
Because that would also mean that the iteration is reported multiple times to JUnit and with the same unique id.
I doubt this would really be a good idea.

If you for example have the `@Retry` extension in mind, what it  does is different, because it has a feature method interceptor that simply calls `invocation.proceed()` multiple times, collects the exceptions and then decides to either rethrow them combined if too many retries or let the feature succeed if no exception happend and only after that the listener that sets the result in the `IterationNode` kicks in. So if you mean this tactic, then it does not object to what I did.

And besides from where to store the result, these open points remain:
- should an unrolled feature signal the state of its iterations
  - imho yes, if you collapse all features and all is green, this sends a wrong signal and if you look in some report or tool only at the feature it also sends a wrong signal
- should a rolled feature abort after the first failing iteration and only report that
  - imho no, it should run all iterations and report all failures as is documented currently to be able to fix them in bunch, pretty much the same reason why not the whole test execution stops after the first failing test, but continues and reports all failures together

Reply:
ping :-)

Reply:
> imho yes, if you collapse all features and all is green, this sends a wrong signal and if you look in some report or tool only at the feature it also sends a wrong signal

This comes down how smart the Report engine is, there are plans do add data driven specs and maybe sub assertions inside features that would be reported as indivdual nodes. At least I'm against the Test rethrowing the same errors that have already been reported.

> should a rolled feature abort after the first failing iteration and only report that 
I agree

However, at this time I'm still against storing the errors inside the IterationNode, the proper way would be to use the ExecutionListener and the new method on the DynamicTestExecutor.

Reply:
> This comes down how smart the Report engine is

Of course, and I kept to what is present now (exemplary Gradle HTML report, IJ tool window with Gradle delegation, IJ tool windows without Gradle delegation) and these were not smart enough to have a good user experience which was easily fixable on Spock side. :-)

> there are plans do add data driven specs and maybe sub assertions inside features that would be reported as indivdual nodes.

Interesting, is there some specification for these? I'm interested what is planned for the future of Spock.

> At least I'm against the Test rethrowing the same errors that have already been reported.

But it isn't if it is unrolled 😕 It throws a new `"At least one iteration failed"` assertion error.

> However, at this time I'm still against storing the errors inside the IterationNode, the proper way would be to use the ExecutionListener and the new method on the DynamicTestExecutor.

Which would automatically mean (most probably), that the unrolled feature can **not** be marked as failing depending on iteration states. Because if you supply an own `ExecutionListener` it will replace the default one, so unless you get the default one injected into your custom listener or something, you cannot listen to the result of the iterations and at the same time have the iterations reported as own nodes, because that requires the default listener to be called accordingly.

Reply:
> Like this, but ignore the names, for this run there were **no** `@PendingFeature` annotations present:
> ![grafik](https://user-images.githubusercontent.com/325196/75081708-717bd500-5510-11ea-8d0a-bdee3185d039.png)

I don't quite get the output. So the result shown here is actually without the `@PendingFeature` extension being active? The results look a bit weird then.

Just to make sure I read this correctly:
* green checkmark: passed
* grey stop sign: skipped/aborted
* yellow x: assertion failed
* red !: failed with exception

* Skipped, Passed -> Skipped?
* exception, exception-> assertion?
* assertion, exception-> assertion?
* Passed, exception-> assertion?
* Skipped, exception-> assertion?
...

Reply:
Exactly, no `@PendingFeature` as it is broken anyway with unrolled features until my PR 1103 makes it in, just ignore them in the name.

The features with `with @Unroll` in the name are unrolled.
The features without `with @Unroll` in the name are rolled up, so after my PR 1101 makes it in, you will only see the feature, not the iterations.

`Skipped, Passed -> Skipped?` looks wrong indeed, but I think that was a display quirk in IJ.

spock-specs/src/test/groovy/org/spockframework/smoke/parameterization/ParameterizedFeatureNodeStatuses.groovy has tests for all combinations
https://github.com/spockframework/spock/pull/1100/files#diff-ab117e2ce273af51aa0ea85f31bc8173R450 is the test that verifies `Skipped, Passed -> Passed` for unrolled feature.

Regarding the others, the logic is:
- all iterations aborted, feature is aborted, no matter whether unrolled or uprolled
- if at least one iteration is not aborted, feature is never aborted, no matter whether unrolled or uprolled

- when unrolled, the feature is never red, because the single iterations are reported for details, it is always a `SpockAssertionError` with `At least one iteration failed` message.
- when uprolled and one iteration failed, the error is simply rethrown
- when uprolled and multiple iterations failed, the failures are aggregated in a `MultipleFailuresError` that is then thrown

Reply:
Ok, as discussed off-site, removed the result aggregation for unrolled features